src/Data/Fin/Properties.agda

```diff
@@ -570,6 +587,76 @@ combine-remQuot {suc n} k i with splitAt k i | P.inspect (splitAt k) i
+  k ℕ.* toℕ (suc x) ℕ.+ toℕ y    ∎
+  where
+    open ≡-Reasoning
+    open import Data.Nat.Solver
+    open +-*-Solver
+
+combine-injectiveˡ : ∀ {n k} (x y : Fin n) (z : Fin k) → combine x z ≡ combine y z → x ≡ y
+combine-injectiveˡ {n} {k@(suc _)} x y z combine[x,z]≡combine[y,z] = toℕ-injective (ℕₚ.*-cancelˡ-≡ k (ℕₚ.+-cancelʳ-≡ (k ℕ.* toℕ x) (k ℕ.* toℕ y) (begin
+  k ℕ.* toℕ x ℕ.+ toℕ z      ≡˘⟨ toℕ-combine x z ⟩
+  toℕ (combine x z)          ≡⟨ cong toℕ combine[x,z]≡combine[y,z] ⟩
+  toℕ (combine y z)          ≡⟨ toℕ-combine y z ⟩
+  k ℕ.* toℕ y ℕ.+ toℕ z      ∎)))
+  where open ≡-Reasoning
+
+combine-injectiveʳ : ∀ {n k} (x : Fin n) (y z : Fin k) → combine x y ≡ combine x z → y ≡ z
+combine-injectiveʳ {n} {k} x y z combine[x,z]≡combine[y,z] = toℕ-injective (ℕₚ.+-cancelˡ-≡ (k ℕ.* toℕ x) (begin
+  k ℕ.* toℕ x ℕ.+ toℕ y ≡˘⟨ toℕ-combine x y ⟩
+  toℕ (combine x y)     ≡⟨ cong toℕ combine[x,z]≡combine[y,z] ⟩
+  toℕ (combine x z)     ≡⟨ toℕ-combine x z ⟩
+  k ℕ.* toℕ x ℕ.+ toℕ z ∎))
+  where open ≡-Reasoning
+
+combine-injective : ∀ {n k} (x : Fin n) (y : Fin k) (w : Fin n) (z : Fin k) → combine x y ≡ combine w z → x ≡ w × y ≡ z
+combine-injective x y w z combine[x,y]≡combine[w,z] = lemma₂ x y w z combine[x,y]≡combine[w,z] , lemma₃ x y w z combine[x,y]≡combine[w,z] where
+  lemma₁ : ∀ {n k} (x : Fin n) (y : Fin k) (w : Fin n) (z : Fin k) → x < w → combine x y < combine w z
+  lemma₁ {n} {k} x y w z x<w = toℕ-cancel-< (begin-strict
+    toℕ (combine x y)      ≡⟨ toℕ-combine x y ⟩
+    k ℕ.* toℕ x ℕ.+ toℕ y  <⟨ ℕₚ.+-monoʳ-< (k ℕ.* toℕ x) (toℕ<n y) ⟩
+    k ℕ.* toℕ x ℕ.+ k      ≡⟨ solve 2 (λ k x → k :* x :+ k := k :* (con 1 :+ x)) refl k (toℕ x) ⟩
+    k ℕ.* suc (toℕ x)      ≤⟨ ℕₚ.*-monoʳ-≤ k (toℕ-mono-< x<w) ⟩
+    k ℕ.* toℕ w            ≤⟨ ℕₚ.m≤m+n (k ℕ.* toℕ w) (toℕ z) ⟩
+    k ℕ.* toℕ w ℕ.+ toℕ z  ≡˘⟨ toℕ-combine w z ⟩
+    toℕ (combine w z)      ∎)
+    where
+      open ℕₚ.≤-Reasoning
+      open import Data.Nat.Solver
+      open +-*-Solver
+  lemma₂ : ∀ {n k} (x : Fin n) (y : Fin k) (w : Fin n) (z : Fin k) → combine x y ≡ combine w z → x ≡ w
+  lemma₂ x y w z combine[x,y]≡combine[w,z] with <-cmp x w
+  ... | tri< x<w _ _ = contradiction combine[x,y]≡combine[w,z] (<⇒≢ (lemma₁ x y w z x<w))
+  ... | tri≈ _ x≡w _ = x≡w
+  ... | tri> _ _ x>w = contradiction (sym combine[x,y]≡combine[w,z]) (<⇒≢ (lemma₁ w z x y x>w))
+  lemma₃ : ∀ {n k} (x : Fin n) (y : Fin k) (w : Fin n) (z : Fin k) → combine x y ≡ combine w z → y ≡ z
+  lemma₃ x y w z combine[x,y]≡combine[w,z] = combine-injectiveʳ x y z (begin
+    combine x y ≡⟨ combine[x,y]≡combine[w,z] ⟩
+    combine w z ≡˘⟨ cong (λ h → combine h z) (lemma₂ x y w z combine[x,y]≡combine[w,z]) ⟩
+    combine x z ∎)
+    where open ≡-Reasoning
+
+combine-surjective : ∀ {n k} (x : Fin (n ℕ.* k)) → Σ[ y ∈ Fin n ] Σ[ z ∈ Fin k ] combine y z ≡ x
+combine-surjective {n} {k} x with remQuot {n} k x | P.inspect (remQuot {n} k) x
+... | y , z | P.[ eq ] = y , z , (begin
```

Comment:
No need for brackets around the equational reasoning proof.

Reply:
It fails to parse otherwise.